internal/db/lint/lint.go

```diff
@@ -84,6 +84,16 @@ func printResultJSON(result []Result, minLevel LintLevel, stdout io.Writer) erro
 	if err := enc.Encode(filtered); err != nil {
 		return errors.Errorf("failed to print result json: %w", err)
 	}
+	// Return an error if fail-on is set and a lint above the level is found
+	if failOn != -1 {
+		for _, r := range filtered {
+			for _, issue := range r.Issues {
+				if toEnum(issue.Level) >= failOn {
+					return errors.New(fmt.Sprintf("fail-on is set to %s, non-zero exit", AllowedLevels[failOn]))
+				}
+			}
+		}
+	}
```

Comment:
I would move this logic to `Run` method so it's separate from printing json result.